### PR TITLE
[WIP] PTQ for QAT initialization for the classification sample

### PR DESCRIPTION
**Pull request by daniil-lyakhov**

### CODE REVIEW

File: tests/qat_after_ptq/torch/test_qat_classification.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+
+def accuracy_drop_is_acceptable(acc_drop: float) -> bool:
+    """
+    Returns True in case acc_drop is less than 1 precent.
+    """
+    return acc_drop < 1.0
+
+
+def catch_nans_wrapper(fn):
```
Comment (ljaljushkin) [inline tests/qat_after_ptq/torch/test_qat_classification.py:183]: why not to use it as a test that the training is correct ?

File: tests/qat_after_ptq/torch/test_qat_classification.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    """
+    model, _ = prepare_model_for_execution(model, config)
+    if config.distributed:
+        broadcast_initialized_parameters(model)
+    model_name = config["model"]
+    train_criterion_fn = inception_criterion_fn if "inception" in model_name else default_criterion_fn
+    params_to_optimize = get_parameter_groups(model, config)
+    optimizer, lr_scheduler = make_optimizer(params_to_optimize, config)
+    best_acc1 = 0
+    logger.info("Qantization aware training pipeline starts.")
```
Comment (ljaljushkin) [inline tests/qat_after_ptq/torch/test_qat_classification.py:210]:
```suggestion
    logger.info("Quantization aware training pipeline starts.")
```

Comment (ljaljushkin) [reply tests/qat_after_ptq/torch/test_qat_classification.py:210]: and the same in other samples

File: tests/qat_after_ptq/torch/test_qat_object_detection.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        config.batch_size //= config.ngpus_per_node
+        config.workers //= config.ngpus_per_node
+
+    acc_drop = train(quantized_model, config, criterion, datasets, original_metric, get_mocked_compression_ctrl())
+    assert accuracy_drop_is_acceptable(acc_drop)
+
+
+def accuracy_drop_is_acceptable(acc_drop: float) -> bool:
+    """
+    Returns True in case acc_drop is less than 1 precent.
```
Comment (ljaljushkin) [inline tests/qat_after_ptq/torch/test_qat_object_detection.py:191]:
```suggestion
    Returns True in case acc_drop is less than 1 percent.
```

**Review by ljaljushkin — CHANGES_REQUESTED** [review]
From the table with results, I can see that training is called for classification and if PTQ accuracy is not good.
What about doing training anyway, but for small number of steps with minimal learning rate and check that there are not zero gradients, loss is changing and accuracy is almost the same?

File: tests/qat_after_ptq/torch/test_qat_classification.py (hunk 1 above)
Comment (daniil-lyakhov) [reply tests/qat_after_ptq/torch/test_qat_classification.py]: I made a separate check, please see

File: tests/qat_after_ptq/torch/test_qat_classification.py (hunk 2 above)
Comment (daniil-lyakhov) [reply tests/qat_after_ptq/torch/test_qat_classification.py]: Done, thanks!

File: tests/qat_after_ptq/torch/test_qat_object_detection.py (hunk 3 above)
Comment (daniil-lyakhov) [reply tests/qat_after_ptq/torch/test_qat_object_detection.py:195]: Done

**Comment by ljaljushkin** [comment]
@daniil-lyakhov do you have a build in CI with these tests?

File: tests/torch/qat_after_ptq/test_qat_classification.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        original_accuracy,
+        get_mocked_compression_ctrl(),
+    )
+    assert accuracy_drop_is_acceptable(acc_drop)
+    check_training_correctness(config, model, datasets, criterion, train_criterion_fn)
+    logger.info("Done!")
+
+
+def accuracy_drop_is_acceptable(acc_drop: float) -> bool:
+    """
```
Comment (AlexanderDokuchaev) [inline tests/torch/qat_after_ptq/test_qat_classification.py:188]: You add docstring for only one the simplest function :satisfied:

File: tests/torch/qat_after_ptq/conftest.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture(scope="session", name="data_dir")
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/qat_after_ptq/conftest.py]: Done

File: tests/torch/qat_after_ptq/helpers.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        return None
+
+    if mode == "symmetric":
+        return QuantizationScheme.SYMMETRIC
+    if mode == "asymmetric":
+        return QuantizationScheme.ASYMMETRIC
+    raise RuntimeError(f"Unknown quantization mode: {mode}")
+
+
+def convert_quantization_params(conf: Optional[dict]) -> QuantizationParameters:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/qat_after_ptq/helpers.py]: Done

File: tests/torch/qat_after_ptq/test_qat_classification.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        val_data_loader=val_data_loader,
+        train_sampler=train_sampler,
+        calibration_dataset=calibration_dataset,
+    )
+
+
+@pytest.mark.weekly
+def test_compression_training(quantization_config_path: Path, data_dir: str, mocker):
+    if "imagenet" not in quantization_config_path.stem:
+        pytest.skip("Test works only with imagenet models by far.")
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/qat_after_ptq/test_qat_classification.py]: It works in the same way, but nevertheless I refactored it

File: tests/torch/qat_after_ptq/test_qat_classification.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    if config.model == "mobilenet_v3_small":
+        # Use default range initializer for mobilenet_v3_small
+        # as due to PTQ advantages it works better for the model.
+        del config.nncf_config["compression"]["initializer"]["range"]
+        del config["compression"]["initializer"]["range"]
+
+    start_worker(main_worker, config)
+
+
+def main_worker(current_gpu: int, config: SampleConfig):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/qat_after_ptq/test_qat_classification.py]: Done

File: tests/torch/qat_after_ptq/test_qat_classification.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+
+def check_training_correctness(
+    config: SampleConfig,
+    model: torch.nn.Module,
+    datasets: DatasetSet,
+    criterion: torch.nn.Module,
+    train_criterion_fn: callable,
+):
+    logger.info("Check model is trainable...")
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/qat_after_ptq/test_qat_classification.py:194]: Done

**Review by alexsu52 — COMMENTED** [review]
tests/torch/qat_after_ptq -> tests/torch/qat

**Review by ljaljushkin — APPROVED** [review]

**Comment by AlexanderDokuchaev** [comment]
torch_weekly/31 - failed

**Comment by daniil-lyakhov** [comment]
> torch_weekly/31 - failed

torch_weekly/38 - in progress

**Review by alexsu52 — APPROVED** [review]
LGTM

**Comment by daniil-lyakhov** [comment]
Build torch_weekly/42/ is successfull
